text/3660-crates-io-crate-deletions.md

```diff
@@ -0,0 +1,130 @@
+Situations like the [`everything` package on npm](https://uncenter.dev/posts/npm-install-everything/) require manual intervention anyway, so we decided to keep the restrictions strict.
+
+> What is the impact of not doing this?
+
+The proposed design is based on the current informal rules that the crates.io team uses to decide whether to delete a crate or version. If we don't implement this proposal, we will continue to rely on the crates.io team to handle these requests manually, which is time-consuming and error-prone.
+
+# Prior art
+
+## npm
+
+The main inspiration for this proposal comes from the npm registry, which has a similar policy for deleting packages and versions:
+
+- https://docs.npmjs.com/policies/unpublish
+- https://docs.npmjs.com/unpublishing-packages-from-the-registry
+
+The npm registry started with a more permissive policy, but had to tighten it over time. 
+It started out with a policy that allowed package owners to delete their packages at any time, but this led to a number of issues, [such as packages being deleted that were depended on by other packages](https://en.wikipedia.org/wiki/Npm_left-pad_incident).
+Their policy was later changed to require that packages can only be deleted within 72 hours of being published, and then [changed again in January 2020](https://blog.npmjs.org/post/190553543620/changes-to-npmunpublish-policy-january-2020) to allow deletions outside the 72-hour window under certain conditions.
+
+
+## PyPI
+
+The Python Package Index (PyPI) still allows package owners to delete their packages (or a subset of released files) at any time.
+A member of the PyPI team has proposed to [stop allowing deleting things from PyPI](https://discuss.python.org/t/stop-allowing-deleting-things-from-pypi/17227) due to the same issues that the npm registry faced. The most current proposed ruleset can be found [here](https://discuss.python.org/t/stop-allowing-deleting-things-from-pypi/17227/71).
+
+Their proposal is also inspired by the npm registry policy, but notably does not include a reverse dependency restriction. It seems that PyPI might not currently be tracking dependencies between packages, which would make it harder for them to implement such a restriction.
+
+## Others
+
+<https://discuss.python.org/t/stop-allowing-deleting-things-from-pypi/17227/59> contains a list of other package registries and their deletion policies.
+
+
+# Unresolved questions
+
+## Should names of deleted crates be blocked so that they can't be re-used?
+
+The reason for this would be to prevent someone else from re-publishing a crate with the same name, which could lead to potential security issues.
+Due to the restrictions on the number of downloads and reverse dependencies, this seems like a low risk though.
+The advantage of allowing others to re-use such names is that it allows name-squatted/placeholder crates to be released back to the community without the crates.io team having to manually intervene.
```

Comment:
Are we expecting to at least clearly flag in the UI or index that the crate existed under some other author / version series? I could see this being very confusing when looking at e.g. crater results or other cases where we try to pull old crates.

Reply:
A marker in the index is one of several topics discussed at https://github.com/rust-lang/rfcs/pull/3660#discussion_r1647537957

Reply:
Right, I saw that, but I guess I interpreted that as mostly being when the version isn't reused - it's not obvious (lacking dates?) that we have a mechanism for identifying a crate after this RFC other than by hash of it's contents, since the same name and version can refer to ~infinitely many different hashes afaict.